### PR TITLE
Update curve25519 dalek

### PR DESCRIPTION
**Pull request by samdealy**
Update `curve25519-dalek` to `4.0.0-pre.5` and make the necessary code changes to acommodate.

### CODE REVIEW

File: rust-toolchain [hunk 1]
```diff
@@ -1 +1 @@
-nightly
+nightly-2022-04-18
```
Comment (remoun) [inline rust-toolchain:1]: Can we use a newer version? Say nightly-2023-01-*

**Review by remoun — DISMISSED** [review]

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,4 @@
 /target/
 **/*.rs.bk
 Cargo.lock
+.idea/*
```
Comment (jcape) [inline .gitignore]: You can ignore this globally:
```
james@jfdi:~ $ cat .gitconfig
[user]
	email = james@mobilecoin.com
	name = James Cape
	signingkey = 42368C417EB57DFE
[core]
	excludesfile = /home/james/.gitignore_global
[init]
	defaultBranch = main
[commit]
	gpgsign = true
```
```
james@jfdi:~ $ cat .gitignore_global
*.code-workspace
.idea/
```

```suggestion
```

File: src/r1cs/mod.rs [hunk 3]
```diff
@@ -1,6 +1,5 @@
-#![cfg_attr(feature = "docs", doc(include = "../../docs/r1cs-docs-example.md"))]
-
-#[cfg_attr(feature = "docs", doc(include = "../../docs/cs-proof.md"))]
+#![doc = include_str!("../../docs/r1cs-docs-example.md")]
```
Comment (jcape) [inline src/r1cs/mod.rs:1]: I think the newline here is actually necessary---`cs-proof.md` is docs for `bulletproofs::r1cs::notes`, whereas `r1cs-docs-example.md` is docs for `bulletproofs::r1cs`.

Without the whitespace, the doc processor will probably treat them as one big blob for one or the other (probably notes).

File: src/r1cs/proof.rs [hunk 4]
```diff
@@ -177,9 +177,12 @@ impl R1CSProof {
...
         let T_6 = CompressedRistretto(read32!());
-        let t_x = Scalar::from_canonical_bytes(read32!()).ok_or(R1CSError::FormatError)?;
-        let t_x_blinding = Scalar::from_canonical_bytes(read32!()).ok_or(R1CSError::FormatError)?;
-        let e_blinding = Scalar::from_canonical_bytes(read32!()).ok_or(R1CSError::FormatError)?;
+        let t_x = Scalar::from_canonical_bytes(read32!());
+        let t_x = Option::from(t_x).ok_or(R1CSError::FormatError)?;
+        let t_x_blinding = Scalar::from_canonical_bytes(read32!());
+        let t_x_blinding = Option::from(t_x_blinding).ok_or(R1CSError::FormatError)?;
+        let e_blinding = Scalar::from_canonical_bytes(read32!());
+        let e_blinding = Option::from(e_blinding).ok_or(R1CSError::FormatError)?;
```
Comment (jcape) [inline src/r1cs/proof.rs:185]: I'd like to get cryptographer eyes on whether this should be CT or not. My instinct is that it doesn't, but it's also parsing blinding factors...

File: src/range_proof/mod.rs [hunk 5]
```diff
@@ -1,13 +1,10 @@
...
+#![doc = include_str!("../../docs/range-proof-protocol.md")]
 
 extern crate alloc;
-#[cfg(feature = "std")]
-extern crate rand;
 
-#[cfg(feature = "std")]
-use self::rand::thread_rng;
 use alloc::vec::Vec;
+use rand::thread_rng;
```
Comment (jcape) [inline src/range_proof/mod.rs]: `rand` needs to continue to be gated on an `std` feature, otherwise we can't use this in the enclave (where `thread_rng()` doesn't exist).

File: src/range_proof/mod.rs [hunk 6]
```diff
@@ -501,6 +498,7 @@ impl RangeProof {
     /// Deserializes the proof from a byte slice.
     ///
     /// Returns an error if the byte slice cannot be parsed into a `RangeProof`.
+    #[allow(clippy::erasing_op)]
```
Comment (jcape) [inline src/range_proof/mod.rs:505]: Why was this required?

**Review by jcape — CHANGES_REQUESTED** [review]
Generally looks good, but we will need to continue to gate `rand` use on `std`, otherwise this will cause linking errors in enclaves.

File: Cargo.toml [hunk 7]
```diff
@@ -18,12 +18,11 @@ description = "A pure-Rust implementation of Bulletproofs using Ristretto"
...
-subtle = { version = "2", default-features = false }
-sha3 = { version = "0.10", default-features = false }
-digest = "0.10"
-rand_core = { version = "0.6", default-features = false, features = ["alloc"] }
-rand = { version = "0.8", default-features = false, optional = true }
+curve25519-dalek = { version = "4.0.0-pre.5", default-features = false, features = ["digest", "rand_core", "serde", "alloc"] }
+subtle = { version = "2.4", default-features = false }
+sha3 = { version = "0.10.6", default-features = false }
+rand_core = { version = "0.6.4", default-features = false, features = ["alloc"] }
+rand = { version = "0.8.5", default-features = false, features = ["std_rng"], optional = true }
```
Comment (isis-mc) [inline Cargo.toml]: Specifying micro versions doesn't change behaviour in semver, unless you want _specifically_ that version and instead you write `version = "=0.10.6"` etc. Same with minor versions for anything above 1.0.0, so saying `subtle = { version = "2" }` is the same functionality as saying `subtle = { version = "2.4" }`. The only line here which changes behaviour is the bump on the pre version of curve.

File: Cargo.toml [hunk 8]
```diff
@@ -35,17 +34,15 @@ clear_on_drop = { version = "0.2", default-features = false }
 hex = "0.4"
 criterion = "0.3"
 bincode = "1"
-rand_chacha = "0.3"
+rand_chacha = "0.3.1"
 
 [features]
 default = ["std"]
-avx2_backend = ["curve25519-dalek/avx2_backend"]
```
Comment (isis-mc) [inline Cargo.toml:42]: Why are we removing the avx2_backend?

File: src/inner_product_proof.rs [hunk 9]
```diff
@@ -400,10 +400,10 @@ impl InnerProductProof {
...
 
         let pos = 2 * lg_n * 32;
-        let a =
-            Scalar::from_canonical_bytes(read32(&slice[pos..])).ok_or(ProofError::FormatError)?;
-        let b = Scalar::from_canonical_bytes(read32(&slice[pos + 32..]))
-            .ok_or(ProofError::FormatError)?;
+        let a = Scalar::from_canonical_bytes(read32(&slice[pos..]));
+        let a = Option::from(a).ok_or(ProofError::FormatError)?;
+        let b = Scalar::from_canonical_bytes(read32(&slice[pos + 32..]));
+        let b = Option::from(b).ok_or(ProofError::FormatError)?;
```
Comment (isis-mc) [inline src/inner_product_proof.rs:406]: `Scalar::from_canonical_bytes()` already returns an `Option`. (Also it's a little weird to `let` assign unmutably to the same variable twice.

File: src/lib.rs [hunk 10]
```diff
@@ -1,6 +1,5 @@
 #![cfg_attr(not(feature = "std"), no_std)]
-#![cfg_attr(feature = "docs", deny(missing_docs))]
-#![cfg_attr(feature = "docs", doc(include = "../README.md"))]
+#![doc = include_str!("../README.md")]
```
Comment (isis-mc) [inline src/lib.rs]: I think we want to keep the `#![cfg_attr(feature = "docs"` on both lines, and the `deny(missing_docs)`

File: src/lib.rs [hunk 11]
```diff
@@ -17,13 +16,13 @@ extern crate serde_derive;
...
+#[doc = include_str!("../docs/notes-intro.md")]
 mod notes {
-    #[cfg_attr(feature = "docs", doc(include = "../docs/notes-ipp.md"))]
+    #[doc = include_str!("../docs/notes-ipp.md")]
     mod inner_product_proof {}
-    #[cfg_attr(feature = "docs", doc(include = "../docs/notes-rp.md"))]
+    #[doc = include_str!("../docs/notes-rp.md")]
     mod range_proof {}
-    #[cfg_attr(feature = "docs", doc(include = "../docs/notes-r1cs.md"))]
+    #[doc = include_str!("../docs/notes-r1cs.md")]
```
Comment (isis-mc) [inline src/lib.rs]: Same as above for all of these.

File: src/r1cs/proof.rs (hunk 4 above)
Comment (isis-mc) [inline src/r1cs/proof.rs:185]: The `Scalar` constructor here is already returning `Option`.

File: src/range_proof/mod.rs [hunk 12]
```diff
@@ -516,12 +514,12 @@ impl RangeProof {
...
-        let t_x_blinding = Scalar::from_canonical_bytes(read32(&slice[5 * 32..]))
-            .ok_or(ProofError::FormatError)?;
-        let e_blinding = Scalar::from_canonical_bytes(read32(&slice[6 * 32..]))
-            .ok_or(ProofError::FormatError)?;
+        let t_x = Scalar::from_canonical_bytes(read32(&slice[4 * 32..]));
+        let t_x = Option::from(t_x).ok_or(ProofError::FormatError)?;
+        let t_x_blinding = Scalar::from_canonical_bytes(read32(&slice[5 * 32..]));
+        let t_x_blinding = Option::from(t_x_blinding).ok_or(ProofError::FormatError)?;
+        let e_blinding = Scalar::from_canonical_bytes(read32(&slice[6 * 32..]));
+        let e_blinding = Option::from(e_blinding).ok_or(ProofError::FormatError)?;
```
Comment (isis-mc) [inline src/range_proof/mod.rs:526]: Already returning Options.

File: src/r1cs/linear_combination.rs [hunk 13]
```diff
@@ -88,7 +88,7 @@ impl Sub<Variable> for Scalar {
 
     fn sub(self, other: Variable) -> Self::Output {
         LinearCombination {
-            terms: vec![(Variable::One(), self), (other, -Scalar::one())],
+            terms: vec![(Variable::One(), self), (other, -Scalar::ONE)],
         }
     }
 }
```
Comment (isis-mc) [inline src/r1cs/linear_combination.rs:94]: Are we planning to pull in the linear proof changes for fixing the frozen heart vulnerability? https://github.com/dalek-cryptography/bulletproofs/pull/361

**Review by isis-mc — COMMENTED** [review]
Some small nits.

File: src/inner_product_proof.rs (hunk 9 above)
Comment (samdealy) [reply src/inner_product_proof.rs:406]: This branch uses curve25519-dalek 4.0.0-rc.5, and this [method](https://docs.rs/curve25519-dalek/4.0.0-rc.1/src/curve25519_dalek/scalar.rs.html#262) returns a CtOption instead of an Optoin (like what earlier versions of this method returned). As such, I implemented this (suboptimal) conversion, but am open to other suggestions?

File: src/r1cs/linear_combination.rs (hunk 13 above)
Comment (samdealy) [reply src/r1cs/linear_combination.rs:94]: I wasn't aware of this change. Should we be planning to pull in this change in this PR, or is it fine (i.e. less urgent) to do so in a follow-up PR?

File: src/r1cs/proof.rs (hunk 4 above)
Comment (samdealy) [reply src/r1cs/proof.rs:185]: See above comment.

File: Cargo.toml (hunk 7 above)
Comment (samdealy) [reply Cargo.toml]: Done.

File: Cargo.toml (hunk 8 above)
Comment (samdealy) [reply Cargo.toml:42]: Because the new version of curve25519-dalek that this branch uses no longer has a `avx2_backend` feature flag (see [here](https://docs.rs/curve25519-dalek/4.0.0-rc.1/curve25519_dalek/#feature-flags)).

File: rust-toolchain (hunk 1 above)
Comment (samdealy) [reply rust-toolchain:1]: I think we can, but would rather do so in a follow-up PR.

File: src/lib.rs (hunk 10 above)
Comment (samdealy) [reply src/lib.rs]: Done.

File: src/lib.rs (hunk 11 above)
Comment (samdealy) [reply src/lib.rs]: Done.

File: src/r1cs/mod.rs (hunk 3 above)
Comment (samdealy) [reply src/r1cs/mod.rs:1]: Done.

File: src/inner_product_proof.rs (hunk 9 above)
Comment (samdealy) [reply src/inner_product_proof.rs:406]: Also, we use shadowing in mobilecoin.git normally in cases when you don't want to rename a variable because they are conceptually the same but are of a slightly different type, but point taken.

File: src/range_proof/mod.rs (hunk 5 above)
Comment (samdealy) [reply src/range_proof/mod.rs]: Done. (I think I did this to make curve25519-dalek compile because it no longer had the std feature or something... just noting this here because I'll have to make some sort of change to curve25519-dalek's usage of the std feature when the time comes).

File: src/range_proof/mod.rs (hunk 6 above)
Comment (samdealy) [reply src/range_proof/mod.rs:505]: Failure to do so causes clippy to fail:
```
error: this operation will always return zero. This is likely not the intended outcome
   --> src/range_proof/mod.rs:515:51
    |
515 |         let A = CompressedRistretto(read32(&slice[0 * 32..]));
    |                                                   ^^^^^^
    |
    = note: `#[deny(clippy::erasing_op)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#erasing_op
```

File: src/range_proof/mod.rs (hunk 12 above)
Comment (samdealy) [reply src/range_proof/mod.rs:526]: See comment above.

**Review by jcape — DISMISSED** [review]

**Review by nick-mobilecoin — DISMISSED** [review]

File: src/r1cs/linear_combination.rs (hunk 13 above)
Comment (samdealy) [reply src/r1cs/linear_combination.rs:94]: Talked with @jcape offline, and we'll pull in changes from upstream in a subsequent PR.

File: src/inner_product_proof.rs (hunk 9 above)
Comment (samdealy) [reply src/inner_product_proof.rs:406]: After discussing with @jcape and @nick-mobilecoin, we'll use this conversion technique.

**Comment by samdealy** [comment]
Had to go back and sign each commit to comply with the base branch's verification policy.

**Review by nick-mobilecoin — APPROVED** [review]
